### DISCUSSION TITLE
How to write a new transformation function in pygimli?

### DISCUSSION THREAD

**Discussion by jinshanmu** [gimli-org/pyGIMLi#648] [Q&A]
I want to write a transformation of porosity to slowness: $s = \frac{1}{{v_m + (v_f - v_m) \cdot \phi}}$ for the TT simulation. This formula could not be written out by the existing functions like pygimli.trans.TransLin() and pygimli.trans.TransPower() individually. So how could I define a tailored transformation function myself?

**Comment by halbmy** [comment]
A very good question pointing out that there is a documentation missing. For an example of writing a transformation function see the SymLog under https://github.com/gimli-org/gimli/blob/c260199832c939f2989710ca570814ca85213d1a/pygimli/core/trans.py#L39
One has to write four methods: a constructor, a forward and reverse transformation, and a derivative, all of them should be easy to implement.

Actually, in the C++ core, there is already a (template) class `TransNest` for combining two transformations, however it is not (yet) available in Python. Then one could easily setup your function as a combination of `TransLin` and `TransPower`. As an example, I will rewrite `TransNest`.

If you want to invert for porosity instead of slowness (the intrinsic property for traveltime), you would have to invert the function, i.e. the forward function is
$$\Phi = \frac{1}{v_f-v_m} \frac{1}{s} - \frac{v_m}{v_f-v_m}$$
the inverse function as written above and the derivative is $\frac{\partial\Phi}{\partial s}$.

**Reply by jinshanmu** [reply]
Thank you for the detailed answer! I'll try to write the function following the example of `TransSymLog`.
